From 2d47b03bd75191a334d1329e8d86190d366b9613 Mon Sep 17 00:00:00 2001
From: Jarno Rajahalme <jarno@covalent.io>
Date: Sun, 31 Jan 2021 19:48:14 -0800
Subject: [PATCH] bazel: Update toolchains to 3.7.2

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>
---
 bazel/repository_locations.bzl | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/bazel/repository_locations.bzl b/bazel/repository_locations.bzl
index e6f1b1afe8..d40f0f4231 100644
--- a/bazel/repository_locations.bzl
+++ b/bazel/repository_locations.bzl
@@ -25,8 +25,8 @@ REPOSITORY_LOCATIONS_SPEC = dict(
         project_name = "bazel-toolchains",
         project_desc = "Bazel toolchain configs for RBE",
         project_url = "https://github.com/bazelbuild/bazel-toolchains",
-        version = "3.7.1",
-        sha256 = "8c9728dc1bb3e8356b344088dfd10038984be74e1c8d6e92dbb05f21cabbb8e4",
+        version = "3.7.2",
+        sha256 = "1caf8584434d3e31be674067996be787cfa511fda2a0f05811131b588886477f",
         strip_prefix = "bazel-toolchains-{version}",
         urls = [
             "https://github.com/bazelbuild/bazel-toolchains/releases/download/{version}/bazel-toolchains-{version}.tar.gz",
-- 
2.24.3 (Apple Git-128)

