From 69d4eb0c25e77defed5cc469a40cf2cba35a61cf Mon Sep 17 00:00:00 2001
From: Tam Mach <sayboras@yahoo.com>
Date: Wed, 22 Jan 2025 18:53:36 +1100
Subject: [PATCH 6/8] patch 6

---
 bazel/foreign_cc/BUILD | 13 +++++++++++++
 1 file changed, 13 insertions(+)

diff --git a/bazel/foreign_cc/BUILD b/bazel/foreign_cc/BUILD
index c903af0465..2dd927e744 100644
--- a/bazel/foreign_cc/BUILD
+++ b/bazel/foreign_cc/BUILD
@@ -11,6 +11,19 @@ envoy_package()
 configure_make(
     name = "liburing",
     configure_in_place = True,
+    env = select({
+        # (sayboras) Ideally, this should be passed as part of environment variables during cross-compilation,
+        # but somehow it doesn't work. So, we have to pass them manually for arm cross-compilation.
+        "//bazel:linux_aarch64": {
+            "CFLAGS": "--target=aarch64-unknown-linux-gnu -fuse-ld=lld",
+            "CPPFLAGS": "--target=aarch64-unknown-linux-gnu -fuse-ld=lld",
+            "LDFLAGS": "-Wl,-S --target=aarch64-unknown-linux-gnu -fuse-ld=lld -Wl,-no-as-needed -Wl,-z,relro,-z,now -lm -l:libstdc++.a -lc",
+            "CC": "/usr/bin/clang-17",
+            "CXX": "/usr/bin/clang-17",
+            "LD": "/usr/bin/lld-17",
+        },
+        "//conditions:default": {},
+    }),
     lib_source = "@com_github_axboe_liburing//:all",
     tags = [
         "nocompdb",
--
2.39.5 (Apple Git-154)

