From 42db1598d81cbccb575c0cd40578f9481a3549e7 Mon Sep 17 00:00:00 2001
From: Tam Mach <sayboras@yahoo.com>
Date: Wed, 14 May 2025 11:27:14 +1000
Subject: [PATCH] build: Fix arm build for liburing

---
 bazel/foreign_cc/BUILD | 14 ++++++++++++++
 1 file changed, 14 insertions(+)

diff --git a/bazel/foreign_cc/BUILD b/bazel/foreign_cc/BUILD
index 2cd885639a..d4b3ea1277 100644
--- a/bazel/foreign_cc/BUILD
+++ b/bazel/foreign_cc/BUILD
@@ -28,6 +28,20 @@ config_setting(
 configure_make(
     name = "liburing",
     configure_in_place = True,
+    env = select({
+        # (sayboras) Ideally, this should be passed as part of environment variables during cross-compilation,
+        # but somehow it doesn't work. So, we have to pass them manually for arm cross-compilation.
+        # The config should be kept in sync with bazel/toolchains/BUILD
+        "//bazel:linux_aarch64": {
+            "CFLAGS": "--target=aarch64-unknown-linux-gnu -fuse-ld=lld-18",
+            "CPPFLAGS": "--target=aarch64-unknown-linux-gnu -fuse-ld=lld-18",
+            "LDFLAGS": "-Wl,-S --target=aarch64-unknown-linux-gnu -fuse-ld=lld-18 -Wl,-no-as-needed -Wl,-z,relro,-z,now -lm -l:libstdc++.a -lc",
+            "CC": "/usr/bin/clang-18",
+            "CXX": "/usr/bin/clang-18",
+            "LD": "/usr/bin/lld-18",
+        },
+        "//conditions:default": {},
+    }),
     lib_source = "@com_github_axboe_liburing//:all",
     tags = [
         "nocompdb",
-- 
2.49.0

