From 73f6930704c7427cdcb3ec72e931835393f74279 Mon Sep 17 00:00:00 2001
From: Tam Mach <sayboras@yahoo.com>
Date: Wed, 14 May 2025 11:27:14 +1000
Subject: [PATCH 6/7] build: Fix arm build for liburing

---
 bazel/foreign_cc/BUILD | 11 +++++++++++
 1 file changed, 11 insertions(+)

diff --git a/bazel/foreign_cc/BUILD b/bazel/foreign_cc/BUILD
index ddc876e95b..d5742e2e08 100644
--- a/bazel/foreign_cc/BUILD
+++ b/bazel/foreign_cc/BUILD
@@ -29,6 +29,17 @@ configure_make(
     name = "liburing",
     configure_in_place = True,
     env = {"ENABLE_SHARED": "0"} | select({
+        # (sayboras) Ideally, this should be passed as part of environment variables during cross-compilation,
+        # but somehow it doesn't work. So, we have to pass them manually for arm cross-compilation.
+        # The config should be kept in sync with bazel/toolchains/BUILD
+        "//bazel:linux_aarch64": {
+            "CFLAGS": "--target=aarch64-unknown-linux-gnu -fuse-ld=lld-18",
+            "CPPFLAGS": "--target=aarch64-unknown-linux-gnu -fuse-ld=lld-18",
+            "LDFLAGS": "-Wl,-S --target=aarch64-unknown-linux-gnu -fuse-ld=lld-18 -Wl,-no-as-needed -Wl,-z,relro,-z,now -lm -l:libstdc++.a -lc",
+            "CC": "/usr/bin/clang-18",
+            "CXX": "/usr/bin/clang-18",
+            "LD": "/usr/bin/lld-18",
+        },
         "//bazel:clang_build": {
             "AR": "$(AR)",
             "RANLIB": "$(AR) -s",
-- 
2.43.0

