From 631fde58a710e67fe39977555c62bf1dcb968b36 Mon Sep 17 00:00:00 2001
From: Jarno Rajahalme <jarno@covalent.io>
Date: Wed, 16 Dec 2020 16:03:23 -0800
Subject: [PATCH] test: Allow use of half-close semantics on TLS connections

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>
---
 test/integration/fake_upstream.cc | 8 ++++++--
 1 file changed, 6 insertions(+), 2 deletions(-)

diff --git a/test/integration/fake_upstream.cc b/test/integration/fake_upstream.cc
index 0b18b9e8c9..7cfceeab68 100644
--- a/test/integration/fake_upstream.cc
+++ b/test/integration/fake_upstream.cc
@@ -591,7 +591,9 @@ FakeUpstream::waitForHttpConnection(Event::Dispatcher& client_dispatcher,
             envoy::config::core::v3::HttpProtocolOptions::ALLOW);
       }
       VERIFY_ASSERTION(connection->initialize());
-      VERIFY_ASSERTION(connection->readDisable(false));
+      if (upstream.read_disable_on_new_connection_) {
+        VERIFY_ASSERTION(connection->readDisable(false));
+      }
       return AssertionSuccess();
     }
   }
@@ -613,7 +615,9 @@ AssertionResult FakeUpstream::waitForRawConnection(FakeRawConnectionPtr& connect
     connection = std::make_unique<FakeRawConnection>(consumeConnection(), timeSystem());
   }
   VERIFY_ASSERTION(connection->initialize());
-  VERIFY_ASSERTION(connection->readDisable(false));
+  if (read_disable_on_new_connection_) {
+    VERIFY_ASSERTION(connection->readDisable(false));
+  }
   VERIFY_ASSERTION(connection->enableHalfClose(enable_half_close_));
   return AssertionSuccess();
 }
-- 
2.24.3 (Apple Git-128)

