*** envoy_bootstrap_v2.json	2020-03-26 14:04:16.000000000 -0700
--- envoy_bootstrap_tmpl.json	2020-04-30 09:46:04.000000000 -0700
***************
*** 260,265 ****
--- 260,279 ----
        },
        {{- end }}
        {
+         "name": "xds-grpc-cilium",
+         "type": "STATIC",
+         "connect_timeout": "{{ .connect_timeout }}",
+         "lb_policy": "ROUND_ROBIN",
+         "hosts": [
+           {
+             "pipe": {
+               "path": "/var/run/cilium/xds.sock"
+             }
+           }
+         ],
+         "http2_protocol_options": { }
+       },
+       {
          "name": "xds-grpc",
          "type": "STRICT_DNS",
          "dns_refresh_rate": "{{ .dns_refresh_rate }}",
